state_machine/src/lib.rs

```diff
@@ -114,42 +110,44 @@ pub struct OverlayedChanges {
 	committed: MemoryState,
 }
 
+const EMPTY_BYTES: [u8; 0] = [];
+
 impl OverlayedChanges {
-	fn code(&self, address: &Address) -> Option<&[u8]> {
-		self.prospective.code(address)
-			.or_else(|| self.committed.code(address))
-			.and_then(|v| if v.is_empty() { None } else { Some(v) })
+	fn code(&self) -> &[u8] {
+		self.prospective.code()
+			.or_else(|| self.committed.code())
+			.unwrap_or_else(|| &EMPTY_BYTES)
 	}
 
-	fn storage(&self, address: &Address, key: &H256) -> Option<&[u8]> {
-		self.prospective.storage(address, key)
-			.or_else(|| self.committed.storage(address, key))
+	fn storage(&self, object: u64, key: &[u8]) -> Option<&[u8]> {
+		self.prospective.storage(object, key)
+			.or_else(|| self.committed.storage(object, key))
 			.and_then(|v| if v.is_empty() { None } else { Some(v) })
 	}
 
 	#[allow(unused)]
-	fn set_code(&mut self, address: Address, code: Vec<u8>) {
-		self.prospective.set_code(address, code);
+	fn set_code(&mut self, code: Vec<u8>) {
+		self.prospective.set_code(code);
 	}
 
-	fn set_storage(&mut self, address: Address, key: H256, val: Vec<u8>) {
-		self.prospective.set_storage(address, key, val);
+	fn set_storage(&mut self, object: u64, key: Vec<u8>, val: Vec<u8>) {
+		self.prospective.set_storage(object, key, val);
 	}
 
 	/// Discard prospective changes to state.
 	pub fn discard_prospective(&mut self) {
-		self.prospective.code.clear();
+		self.prospective.code = None;
 		self.prospective.storage.clear();
 	}
 
 	/// Commit prospective changes to state.
 	pub fn commit_prospective(&mut self) {
-		let code_updates = self.prospective.code.drain()
-			.map(|(addr, code)| Update::Code(addr, code));
+		let code_updates = self.prospective.code.take().into_iter()
+			.map(|code| Update::Code(code));
```

Comment:
`map(Update::Code)`

Reply:
wouldn't that be equivalent to `Update::Code((addr, code))`?